aws/src/main/java/org/apache/iceberg/aws/dynamodb/DynamoDbCatalog.java

```diff
@@ -0,0 +1,621 @@
+    }
+  }
+
+  private void validateTableIdentifier(TableIdentifier identifier) {
+    validateNamespace(identifier.namespace());
+    ValidationException.check(identifier.hasNamespace(),
+        "Table namespace must not be empty: %s", identifier);
+    String tableName = identifier.name();
+    ValidationException.check(!tableName.contains("."),
+        "Table name must not contain dot: %s", tableName);
+  }
+
+  private boolean dynamoDbTableExists(String tableName) {
+    try {
+      dynamo.describeTable(DescribeTableRequest.builder()
+          .tableName(tableName)
+          .build());
+      return true;
+    } catch (ResourceNotFoundException e) {
+      return false;
+    }
+  }
+
+  private void ensureCatalogTableExistsOrCreate() {
+    if (dynamoDbTableExists(awsProperties.dynamoDbTableName())) {
+      return;
+    }
+
+    LOG.info("DynamoDb catalog table {} not found, trying to create", awsProperties.dynamoDbTableName());
+    dynamo.createTable(CreateTableRequest.builder()
+        .tableName(awsProperties.dynamoDbTableName())
+        .keySchema(
+            KeySchemaElement.builder().attributeName(COL_IDENTIFIER).keyType(KeyType.HASH).build(),
+            KeySchemaElement.builder().attributeName(COL_NAMESPACE).keyType(KeyType.RANGE).build())
+        .attributeDefinitions(
+            AttributeDefinition.builder().attributeName(COL_IDENTIFIER).attributeType(ScalarAttributeType.S).build(),
+            AttributeDefinition.builder().attributeName(COL_NAMESPACE).attributeType(ScalarAttributeType.S).build())
+        .globalSecondaryIndexes(GlobalSecondaryIndex.builder()
+            .indexName(GSI_NAMESPACE_IDENTIFIER)
+            .keySchema(
+                KeySchemaElement.builder().attributeName(COL_NAMESPACE).keyType(KeyType.HASH).build(),
+                KeySchemaElement.builder().attributeName(COL_IDENTIFIER).keyType(KeyType.RANGE).build())
+            .projection(Projection.builder().projectionType(ProjectionType.KEYS_ONLY).build())
+            .build())
+        .billingMode(BillingMode.PAY_PER_REQUEST)
+        .build());
+
+    // wait for the dynamo table to complete provisioning, which takes around 10 seconds
+    Tasks.foreach(awsProperties.dynamoDbTableName())
+        .retry(CATALOG_TABLE_CREATION_WAIT_ATTEMPTS_MAX)
+        .throwFailureWhenFinished()
+        .onlyRetryOn(IllegalStateException.class)
+        .run(this::checkTableActive);
+  }
+
+  private void checkTableActive(String tableName) {
+    try {
+      DescribeTableResponse response = dynamo.describeTable(DescribeTableRequest.builder()
+          .tableName(tableName)
+          .build());
+      TableStatus currentStatus = response.table().tableStatus();
+      if (!currentStatus.equals(TableStatus.ACTIVE)) {
+        throw new IllegalStateException(String.format("Dynamo catalog table %s is not active, current status: %s",
+            tableName, currentStatus));
+      }
+    } catch (ResourceNotFoundException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)